pandas/core/internals/construction.py

```diff
@@ -547,20 +557,32 @@ def _ensure_2d(values: np.ndarray) -> np.ndarray:
     return values
 
 
-def _homogenize(data, index: Index, dtype: DtypeObj | None) -> list[ArrayLike]:
+def _homogenize(
+    data, index: Index, dtype: DtypeObj | None
+) -> tuple[list[ArrayLike], list[Any] | None]:
     oindex = None
     homogenized = []
+    # if the original array-like in `data` is a Series, keep track of this Series
+    parents: list[Any] = []
 
     for val in data:
         if isinstance(val, ABCSeries):
+            hval = val
             if dtype is not None:
-                val = val.astype(dtype, copy=False)
-            if val.index is not index:
+                hval = hval.astype(dtype, copy=False)
+                if using_copy_on_write() and hval.values is val.values:
+                    # TODO(CoW) remove when astype() has implemented CoW
+                    refs = [weakref.ref(val._mgr._block)]  # type: ignore[union-attr]
+                    hval._mgr.refs = refs  # type: ignore[union-attr]
+                    hval._mgr.parent = val._mgr  # type: ignore[union-attr]
+            if hval.index is not index:
                 # Forces alignment. No need to copy data since we
                 # are putting it into an ndarray later
-                val = val.reindex(index, copy=False)
-
-            val = val._values
+                hval = hval.reindex(index, copy=False)
+            if hval is val:
+                hval = val.copy(deep=False)
+            homogenized.append(hval._values)
+            parents.append(hval)
```

Comment:
isn't this only relevant if reindex did not make a copy?

Reply:
True this would work, but the following does not work:


```
val = Series([1, 2, 3.0])
df = pd.DataFrame({"a": val}, dtype="int64", index=[0, 1, 2])


assert not np.shares_memory(get_array(df, "a"), val._values)

x = get_array(df, "a")
df.iloc[0, 0] = 100
assert np.shares_memory(get_array(df, "a"), x)
```

In this scenario astype makes a copy while reindex does not. Hence, hval has a ref to the object created from astype which is inherited by the constructed element that is our result. Consequentially, iloc triggers an unnecessary copy

Reply:
Hmm, yes, if you have both a reindex and astype, that makes things a bit more complicated ..

Reply:
Updated for this by checking in each step if an actual copy was made, and then if any of the steps did make a copy, we don't add the Series to the parents (thus, also if the first astype step copied, and then the next reindex step did not copy, we wouldn't add the parent, and not trigger a CoW unnecessarily)

And added a test that covers your example above (that was failing before the fix)